libs/s25main/ogl/glSmartBitmap.cpp

```diff
@@ -253,35 +275,39 @@ void glSmartBitmap::drawPercent(DrawPoint drawPt, unsigned percent, unsigned col
             return;
     }
 
-    // nothing to draw?
-    if(!percent)
-        return;
-    RTTR_Assert(percent <= 100);
-
-    const float partDrawn = percent / 100.f;
     std::array<Point<GLfloat>, 8> vertices, curTexCoords;
     std::array<GL_RGBAColor, 8> colors;
 
+    auto drawPt = dstArea.getOrigin();
     drawPt -= origin_;
-    vertices[2] = Point<GLfloat>(drawPt) + size_;
+    vertices[2] = Point<GLfloat>(dstArea.getEndPt() - origin_); // destination bottom
 
     vertices[0].x = vertices[1].x = GLfloat(drawPt.x);
     vertices[3].x = vertices[2].x;
 
-    vertices[0].y = vertices[3].y = GLfloat(drawPt.y + size_.y * (1.f - partDrawn));
-    vertices[1].y = vertices[2].y;
+    vertices[0].y = vertices[3].y = GLfloat(drawPt.y); // destination top
+    vertices[1].y = vertices[2].y;                     // destination bottom
 
     colors[0].r = GetRed(color);
     colors[0].g = GetGreen(color);
     colors[0].b = GetBlue(color);
     colors[0].a = GetAlpha(color);
     colors[3] = colors[2] = colors[1] = colors[0];
 
+    // vertical coords
     curTexCoords[0] = texCoords[0];
     curTexCoords[1] = texCoords[1];
     curTexCoords[2] = texCoords[2];
     curTexCoords[3] = texCoords[3];
-    curTexCoords[0].y = curTexCoords[3].y = curTexCoords[1].y - (curTexCoords[1].y - curTexCoords[0].y) * partDrawn;
+    curTexCoords[0].y = curTexCoords[3].y = helpers::lerp(
+      texCoords[0].y, texCoords[1].y, helpers::inverseLerp(.0f, float(size_.y), float(srcArea.getOrigin().y)));
+    curTexCoords[1].y = curTexCoords[2].y = helpers::lerp(
+      texCoords[0].y, texCoords[1].y, helpers::inverseLerp(.0f, float(size_.y), float(srcArea.getEndPt().y)));
+    // horizontal coords
+    curTexCoords[0].x = curTexCoords[1].x = helpers::lerp(
+      texCoords[0].x, texCoords[3].x, helpers::inverseLerp(.0f, float(size_.x), float(srcArea.getOrigin().x)));
+    curTexCoords[2].x = curTexCoords[3].x = helpers::lerp(
+      texCoords[0].x, texCoords[3].x, helpers::inverseLerp(.0f, float(size_.x), float(srcArea.getEndPt().x)));
```

Comment:
I have trouble understanding this lerp of inverseLerp. Can you introduce variables for the inverseLerp please? Giving those values names that might be enough to understand what they are. If not a comment what the inverseLerp caluclates will do (in addition to the variables)
Especially as the difference in the inverseLerps is easy to miss.

Reply:
I don't quite remember RTTR's internals but it appears that texCoords and srcArea were using coordinate systems - the lerp(inverseLerp()) combo was for remapping between their ranges.

I've simplified the code a bit so perhaps it's more readable what's going on.